### PR TITLE
[DO NOT MERGE] [PDS-146088] Test infinite loop hypothesis

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- PDS-146088

**Implementation Description (bullets)**:
- Copy the generic object pool code to Atlas.
- Introduce debug logs at key places within the client pool.

**Testing (What was existing testing like?  What have you done to improve it?)**: existing tests should be enough?

**Concerns (what feedback would you like?)**:
- Does this break anything?
- Is licensing done correctly?

**Where should we start reviewing?**: Commit by commit. Commit 2 is a bit large - just confirm it's a copy and then the CHANGELOG bits in there.

**Priority (whenever / two weeks / yesterday)**: this week

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java [hunk 1]
```diff
@@ -348,6 +349,11 @@ private NonEvictionLoggingEvictionPolicy(EvictionPolicy<T> delegate) {
 
         @Override
         public boolean evict(EvictionConfig config, PooledObject<T> underTest, int idleCount) {
+            if (log.isDebugEnabled()) {
+                log.debug("Beginning an eviction attempt from the Cassandra client pool",
+                        SafeArg.of("underTestState", underTest.getState()),
+                        SafeArg.of("idleState", underTest.getIdleTimeMillis()));
+            }
             boolean delegateResult = delegate.evict(config, underTest, idleCount);
```
Comment (Jolyon-S) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java:357]: Do we want to wrap with `TimedRunner`?

File: atlasdb-cassandra/src/main/java/org/apache/commons/pool2/impl/CommonsPoolGenericObjectPool.java [hunk 2]
```diff
@@ -774,14 +777,31 @@ public void evict() throws Exception {
                     } catch (final NoSuchElementException nsee) {
                         // Object was borrowed in another thread
                         // Don't count this as an eviction test so reduce i;
+                        // CHANGELOG: Additional logging
+                        if (log.isDebugEnabled()) {
+                            log.debug("Not counting an eviction candidate, because we believe it was borrowed in "
```
Comment (Jolyon-S) [inline atlasdb-cassandra/src/main/java/org/apache/commons/pool2/impl/CommonsPoolGenericObjectPool.java:784]: I suggest that we also log:
* `i`
* `m`
* some sort of UUID per eviction run. Then we can group logs by eviction run to confirm that it is indeed the same run that is looping indefinitely.

File: atlasdb-cassandra/src/main/java/org/apache/commons/pool2/impl/CommonsPoolGenericObjectPool.java [hunk 3]
```diff
@@ -774,14 +777,31 @@ public void evict() throws Exception {
...
 
+                    // CHANGELOG: Declaration of a variable to check the object's state before starting eviction.
+                    PooledObjectState preTestState = underTest.getState();
+
                     if (!underTest.startEvictionTest()) {
                         // Object was borrowed in another thread
                         // Don't count this as an eviction test so reduce i;
+                        // CHANGELOG: Additional logging
+                        if (log.isDebugEnabled()) {
+                            log.debug("Not counting an eviction candidate, because we weren't able to start an "
```
Comment (Jolyon-S) [inline atlasdb-cassandra/src/main/java/org/apache/commons/pool2/impl/CommonsPoolGenericObjectPool.java:800]: As above

File: atlasdb-cassandra/src/main/java/org/apache/commons/pool2/impl/CommonsPoolGenericObjectPool.java (hunk 2 above)
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/org/apache/commons/pool2/impl/CommonsPoolGenericObjectPool.java:788]: Makes sense.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java:357]: We could, but I think it's probably not needed. The delegate impl looks like this:

```
    @Override
    public boolean evict(final EvictionConfig config, final PooledObject<T> underTest,
            final int idleCount) {

        if ((config.getIdleSoftEvictTime() < underTest.getIdleTimeMillis() &&
                config.getMinIdle() < idleCount) ||
                config.getIdleEvictTime() < underTest.getIdleTimeMillis()) {
            return true;
        }
        return false;
    }
```

so this should be a bunch of memory reads. Actually this starting log block isn't probably necessary in fact.

**Comment by jeremyk-91** [comment]
closing in favour of https://github.com/palantir/atlasdb/pull/5503 - this turned out to be correct! but there's a simpler fix

**Pull request closed without merging**